File: SuperGrid/C#/StateControl.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Text;
using System.Windows.Forms;

namespace SuperGridDemo
{
    public partial class StateControl : Control
    {
        #region Events

        #region StateChanged

        /// <summary>
        /// Occurs when the control state has changed
        /// </summary>
        [Description("Occurs when the control state has changed.")]
        public event EventHandler<StateChangedEventArgs> StateChanged;

        #endregion

        #endregion

        #region Private variables

        private MyState[] _States;

        private MyState _HitItem;
        private MyState _LastHitItem;
        private MyState _MouseDownItem;

        private string _Value;
        private bool _NeedRecalc;
        private Size _PreferredSize;

        #endregion

        public StateControl()
        {
            InitializeComponent();
        }

        #region Public properties

        #region Value

        /// <summary>
        /// Gets or sets the comma delimited state string.
        /// </summary>
        protected string Value
        {
            get
            {
                if (_Value == null)
                    _Value = GetStates();

                return (_Value);
            }

            set
            {
                if ((value != null && value.Equals(_Value) == false) ||
                    (value == null && value != _Value))
                {
                    _Value = value;

                    SetStates(value);
                }
            }
        }

        #region GetStates

        /// <summary>
        /// Gets the states "string" value.
        /// </summary>
        /// <returns></returns>
        private string GetStates()
        {
            StringBuilder sb = new StringBuilder();

            if (_States != null)
            {
                foreach (MyState state in _States)
                {
                    sb.Append(state.ColorIndex);
                    sb.Append(",");
                }

                if (sb.Length > 0)
                    sb.Length--;
            }

            return (sb.ToString());
        }

        #endregion

        #region SetStates

        /// <summary>
        /// Sets the states "string" value
        /// </summary>
        /// <param name="value"></param>
        private void SetStates(string value)
        {
            if (value == null)
            {
                _States = null;
            }
            else
            {
                string[] states = value.Split(',');
                
                _States = new MyState[states.Length];

                int i = 0;
                foreach (string s in states)
                {
                    int n;
                    int.TryParse(s, out n);

                    _States[i++] = new MyState(n);
                }
            }

            _NeedRecalc = true;
        }

        #endregion

        #endregion

        #endregion

        #region OnPaint

        /// <summary>
        /// Handles paint event requests
        /// </summary>
        /// <param name="e"></param>
        protected override void OnPaint(PaintEventArgs e)
        {
            if (_States != null)
            {
                if (_NeedRecalc == true)
                    RecalcLayout();

                Graphics g = e.Graphics;
                g.SmoothingMode = SmoothingMode.AntiAlias;

                foreach (MyState state in _States)
                {
                    Rectangle r = state.Bounds;

                    if (_HitItem == state)
                        r.Inflate(1, 1);

                    int index = state.ColorIndex;

                    if ((uint) index >= imageList1.Images.Count)
                        index = 0;

                    g.DrawImageUnscaledAndClipped(imageList1.Images[index], r);
                }
            }
        }

        #endregion

        #region RecalcLayout

        /// <summary>
        /// Recalculates the states layout
        /// </summary>
        private void RecalcLayout()
        {
            _PreferredSize = Bounds.Size;

            if (_States != null)
            {
                Rectangle r = Bounds;
                r.Inflate(-2, -2);
                r.Width = 16;

                r.Y += (r.Height - 16)/2;
                r.Height = 16;

                foreach (MyState state in _States)
                {
                    state.Bounds = r;

                    r.X += 18;
                }

                _PreferredSize.Width = r.X - Bounds.X;
            }

            _NeedRecalc = false;
        }

        #endregion

        #region GetPreferredSize

        /// <summary>
        /// Gets the control's preferred size
        /// </summary>
        /// <param name="proposedSize"></param>
        /// <returns></returns>
        public override Size GetPreferredSize(Size proposedSize)
        {
            if (_NeedRecalc == true)
                RecalcLayout();

            return (_PreferredSize);
        }

        #endregion

        #region Mouse support

        #region OnMouseMove

        /// <summary>
        /// Handles MouseMove events
        /// </summary>
        /// <param name="e"></param>
        protected override void OnMouseMove(MouseEventArgs e)
        {
            base.OnMouseMove(e);

            _HitItem = GetHitItem(e.Location);

            if (_HitItem != _LastHitItem)
            {
                _LastHitItem = _HitItem;

                Invalidate();
            }
        }

        #region GetHitItem

        /// <summary>
        /// Gets the item currently under the mouse
        /// </summary>
        /// <param name="pt"></param>
        /// <returns></returns>
        private MyState GetHitItem(Point pt)
        {
            if (_States != null)
            {
                foreach (MyState state in _States)
                {
                    if (state.Bounds.Contains(pt) == true)
                        return (state);
                }
            }

            return (null);
        }

        #endregion

        #endregion

        #region OnMouseDown

        /// <summary>
        /// Handles MouseDown events
        /// </summary>
        /// <param name="e"></param>
        protected override void OnMouseDown(MouseEventArgs e)
        {
            base.OnMouseDown(e);

            _MouseDownItem = _HitItem;
        }

        #endregion

        #region OnMouseUp

        /// <summary>
        /// Handles MouseUp events
        /// </summary>
        /// <param name="e"></param>
        protected override void OnMouseUp(MouseEventArgs e)
        {
            // Change the state colorIndex
            // if the user clicked on an item.

            if (_MouseDownItem != null && _MouseDownItem == _HitItem)
            {
                _MouseDownItem.ColorIndex =
                    (_MouseDownItem.ColorIndex + 1) % imageList1.Images.Count;

                OnValueChanged();
            }

            _MouseDownItem = null;

            base.OnMouseUp(e);
        }

        #region OnValueChanged

        /// <summary>
        /// OnValueChanged
        /// </summary>
        private void OnValueChanged()
        {
            string oldValue = Value;

            _Value = null;

            if (StateChanged != null)
                StateChanged(this, new StateChangedEventArgs(oldValue, Value));
        }

        #endregion

        #endregion

        #endregion
    }

    #region StateChangedEventArgs

    /// <summary>
    /// StateChangedEventArgs
    /// </summary>
    public class StateChangedEventArgs : EventArgs
    {
        #region Private variables

        private string _OldValue;
        private string _NewValue;

        #endregion

        ///<summary>
        /// StateChangedEventArgs
        ///</summary>
        ///<param name="oldValue"></param>
        ///<param name="newValue"></param>
        public StateChangedEventArgs(string oldValue, string newValue)
        {
            _OldValue = oldValue;
            _NewValue = newValue;
        }

        #region Public properties

        /// <summary>
        /// Gets the old value
        /// </summary>
        public string OldValue
        {
            get { return (_OldValue); }
        }

        /// <summary>
        /// Gets the new value
        /// </summary>
        public string NewValue
        {
            get { return (_NewValue); }
        }

        #endregion
    }

    #endregion

    #region MyState

    public class MyState
    {
        #region Private variables

        private int _ColorIndex;
        private Rectangle _Bounds;

        #endregion

        public MyState(int colorIndex)
        {
            _ColorIndex = colorIndex;
        }

        #region Public properties

        #region Bounds

        /// <summary>
        /// Gets the state bounding rect
        /// </summary>
        public Rectangle Bounds
        {
            get { return (_Bounds); }
            internal set { _Bounds = value; }
        }

        #endregion

        #region ColorIndex

        /// <summary>
        /// Gets or sets the state color index
        /// </summary>
        public int ColorIndex
        {
            get { return (_ColorIndex); }
            set { _ColorIndex = value; }
        }

        #endregion

        #endregion
    }

    #endregion
}

File: SuperGrid/VB/StateControl.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D
Imports System.Text

Namespace SuperGridDemo
	Partial Public Class StateControl
		Inherits Control
		#Region "Events"

		#Region "StateChanged"

		''' <summary>
		''' Occurs when the control state has changed
		''' </summary>
		<Description("Occurs when the control state has changed.")> _
		Public Event StateChanged As EventHandler(Of StateChangedEventArgs)

		#End Region

		#End Region

		#Region "Private variables"

		Private _States() As MyState

		Private _HitItem As MyState
		Private _LastHitItem As MyState
		Private _MouseDownItem As MyState

		Private _Value As String
		Private _NeedRecalc As Boolean
		Private _PreferredSize As Size

		#End Region

		Public Sub New()
			InitializeComponent()
		End Sub

		#Region "Public properties"

		#Region "Value"

		''' <summary>
		''' Gets or sets the comma delimited state string.
		''' </summary>
		Protected Property Value() As String
			Get
				If _Value Is Nothing Then
					_Value = GetStates()
				End If

				Return (_Value)
			End Get

			Set(ByVal value As String)
				If (value IsNot Nothing AndAlso value.Equals(_Value) = False) OrElse (value Is Nothing AndAlso value <> _Value) Then
					_Value = value

					SetStates(value)
				End If
			End Set
		End Property

		#Region "GetStates"

		''' <summary>
		''' Gets the states "string" value.
		''' </summary>
		''' <returns></returns>
		Private Function GetStates() As String
			Dim sb As New StringBuilder()

			If _States IsNot Nothing Then
				For Each state As MyState In _States
					sb.Append(state.ColorIndex)
					sb.Append(",")
				Next state

				If sb.Length > 0 Then
					sb.Length -= 1
				End If
			End If

			Return (sb.ToString())
		End Function

		#End Region

		#Region "SetStates"

		''' <summary>
		''' Sets the states "string" value
		''' </summary>
		''' <param name="value"></param>
		Private Sub SetStates(ByVal value As String)
			If value Is Nothing Then
				_States = Nothing
			Else
				Dim states() As String = value.Split(","c)

				_States = New MyState(states.Length - 1){}

				Dim i As Integer = 0
				For Each s As String In states
					Dim n As Integer
					Integer.TryParse(s, n)

					_States(i) = New MyState(n)
					i += 1
				Next s
			End If

			_NeedRecalc = True
		End Sub

		#End Region

		#End Region

		#End Region

		#Region "OnPaint"

		''' <summary>
		''' Handles paint event requests
		''' </summary>
		''' <param name="e"></param>
		Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
			If _States IsNot Nothing Then
				If _NeedRecalc = True Then
					RecalcLayout()
				End If

				Dim g As Graphics = e.Graphics
				g.SmoothingMode = SmoothingMode.AntiAlias

				For Each state As MyState In _States
					Dim r As Rectangle = state.Bounds

					If _HitItem Is state Then
						r.Inflate(1, 1)
					End If

					Dim index As Integer = state.ColorIndex

					If CUInt(index) >= imageList1.Images.Count Then
						index = 0
					End If

					g.DrawImageUnscaledAndClipped(imageList1.Images(index), r)
				Next state
			End If
		End Sub

		#End Region

		#Region "RecalcLayout"

		''' <summary>
		''' Recalculates the states layout
		''' </summary>
		Private Sub RecalcLayout()
			_PreferredSize = Bounds.Size

			If _States IsNot Nothing Then
				Dim r As Rectangle = Bounds
				r.Inflate(-2, -2)
				r.Width = 16

				r.Y += (r.Height - 16)\2
				r.Height = 16

				For Each state As MyState In _States
					state.Bounds = r

					r.X += 18
				Next state

				_PreferredSize.Width = r.X - Bounds.X
			End If

			_NeedRecalc = False
		End Sub

		#End Region

		#Region "GetPreferredSize"

		''' <summary>
		''' Gets the control's preferred size
		''' </summary>
		''' <param name="proposedSize"></param>
		''' <returns></returns>
		Public Overrides Function GetPreferredSize(ByVal proposedSize As Size) As Size
			If _NeedRecalc = True Then
				RecalcLayout()
			End If

			Return (_PreferredSize)
		End Function

		#End Region

		#Region "Mouse support"

		#Region "OnMouseMove"

		''' <summary>
		''' Handles MouseMove events
		''' </summary>
		''' <param name="e"></param>
		Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)

			_HitItem = GetHitItem(e.Location)

			If _HitItem IsNot _LastHitItem Then
				_LastHitItem = _HitItem

				Invalidate()
			End If
		End Sub

		#Region "GetHitItem"

		''' <summary>
		''' Gets the item currently under the mouse
		''' </summary>
		''' <param name="pt"></param>
		''' <returns></returns>
		Private Function GetHitItem(ByVal pt As Point) As MyState
			If _States IsNot Nothing Then
				For Each state As MyState In _States
					If state.Bounds.Contains(pt) = True Then
						Return (state)
					End If
				Next state
			End If

			Return (Nothing)
		End Function

		#End Region

		#End Region

		#Region "OnMouseDown"

		''' <summary>
		''' Handles MouseDown events
		''' </summary>
		''' <param name="e"></param>
		Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			MyBase.OnMouseDown(e)

			_MouseDownItem = _HitItem
		End Sub

		#End Region

		#Region "OnMouseUp"

		''' <summary>
		''' Handles MouseUp events
		''' </summary>
		''' <param name="e"></param>
		Protected Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			' Change the state colorIndex
			' if the user clicked on an item.

			If _MouseDownItem IsNot Nothing AndAlso _MouseDownItem Is _HitItem Then
				_MouseDownItem.ColorIndex = (_MouseDownItem.ColorIndex + 1) Mod imageList1.Images.Count

				OnValueChanged()
			End If

			_MouseDownItem = Nothing

			MyBase.OnMouseUp(e)
		End Sub

		#Region "OnValueChanged"

		''' <summary>
		''' OnValueChanged
		''' </summary>
		Private Sub OnValueChanged()
			Dim oldValue As String = Value

			_Value = Nothing

			RaiseEvent StateChanged(Me, New StateChangedEventArgs(oldValue, Value))
		End Sub

		#End Region

		#End Region

		#End Region
	End Class

	#Region "StateChangedEventArgs"

	''' <summary>
	''' StateChangedEventArgs
	''' </summary>
	Public Class StateChangedEventArgs
		Inherits EventArgs
		#Region "Private variables"

		Private _OldValue As String
		Private _NewValue As String

		#End Region

		'''<summary>
		''' StateChangedEventArgs
		'''</summary>
		'''<param name="oldValue"></param>
		'''<param name="newValue"></param>
		Public Sub New(ByVal oldValue As String, ByVal newValue As String)
			_OldValue = oldValue
			_NewValue = newValue
		End Sub

		#Region "Public properties"

		''' <summary>
		''' Gets the old value
		''' </summary>
		Public ReadOnly Property OldValue() As String
			Get
				Return (_OldValue)
			End Get
		End Property

		''' <summary>
		''' Gets the new value
		''' </summary>
		Public ReadOnly Property NewValue() As String
			Get
				Return (_NewValue)
			End Get
		End Property

		#End Region
	End Class

	#End Region

	#Region "MyState"

	Public Class MyState
		#Region "Private variables"

		Private _ColorIndex As Integer
		Private _Bounds As Rectangle

		#End Region

		Public Sub New(ByVal colorIndex As Integer)
			_ColorIndex = colorIndex
		End Sub

		#Region "Public properties"

		#Region "Bounds"

		''' <summary>
		''' Gets the state bounding rect
		''' </summary>
		Public Property Bounds() As Rectangle
			Get
				Return (_Bounds)
			End Get
			Friend Set(ByVal value As Rectangle)
				_Bounds = value
			End Set
		End Property

		#End Region

		#Region "ColorIndex"

		''' <summary>
		''' Gets or sets the state color index
		''' </summary>
		Public Property ColorIndex() As Integer
			Get
				Return (_ColorIndex)
			End Get
			Set(ByVal value As Integer)
				_ColorIndex = value
			End Set
		End Property

		#End Region

		#End Region
	End Class

	#End Region
End Namespace
